src/Microsoft.Tye.Hosting/BuildWatcher.cs

```diff
@@ -0,0 +1,287 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Build.Construction;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Tye.Hosting
+{
+    internal sealed class BuildWatcher : IAsyncDisposable
+    {
+        private CancellationTokenSource? _cancellationTokenSource;
+        private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
+        private readonly ILogger _logger;
+        private Task? _processor;
+        private Channel<BuildRequest>? _queue;
+
+        public BuildWatcher(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public Task StartAsync(string? solutionPath, string workingDirectory)
+        {
+            return WithLockAsync(
+                async () =>
+                {
+                    await ResetAsync();
+
+                    _queue = Channel.CreateUnbounded<BuildRequest>();
+                    _cancellationTokenSource = new CancellationTokenSource();
+                    _processor = Task.Run(() => ProcessTaskQueueAsync(_logger, _queue.Reader, solutionPath, workingDirectory, _cancellationTokenSource.Token));
+                });
+        }
+
+        public Task StopAsync()
+        {
+            return WithLockAsync(ResetAsync);
+        }
+
+        public async Task<int> BuildProjectFileAsync(string projectFilePath)
+        {
+            var request = new BuildRequest(projectFilePath);
+
+            await WithLockAsync(
+                async () =>
+                {
+                    if (_queue == null)
+                    {
+                        throw new InvalidOperationException("The worker is not running.");
+                    }
+
+                    await _queue.Writer.WriteAsync(request);
+                });
+
+            return await request.Task;
+        }
+
+        #region IAsyncDisposable Members
+
+        public async ValueTask DisposeAsync()
+        {
+            await WithLockAsync(ResetAsync);
+
+            _lock.Dispose();
+        }
+
+        #endregion
+
+        private async Task WithLockAsync(Func<Task> action)
+        {
+            await _lock.WaitAsync();
+
+            try
+            {
+                await action();
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        private async Task ResetAsync()
+        {
+            _queue?.Writer.TryComplete();
+            _queue = null;
+
+            _cancellationTokenSource?.Cancel();
+            _cancellationTokenSource?.Dispose();
+            _cancellationTokenSource = null;
```

Comment:
It would be great to reduce the repetitive `null` checks inside this method, similar to what was done for `_processor` in the following lines.

Reply:
Sounds reasonable to me; updated.